### PR TITLE
[DX] Add rule to check not "Tests" namespace outside "tests" directory

### PR DESCRIPTION
**Pull request by dobryy**

### CODE REVIEW

**Comment by dobryy** [comment]
@TomasVotruba 
There're no such cases in the code. I created one to show that new rule works. 
Also the test is failing because the test fixture is in the `tests` directory of the `phpstan-extensions` util.

**Comment by dobryy** [comment]
I can't come up with some solution to the problem

File: utils/phpstan-extensions/src/Rule/CheckNotTestsNamespaceOutsideTestsDirectoryRule.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            return [];
+        }
+
+        if (
+            in_array('Tests', $node->name->parts, true) &&
+            ! strstr($scope->getFileDescription(), '/tests/')
+        ) {
+            $errorMessage = sprintf(self::ERROR_MESSAGE, $node->name->toString(), $scope->getFileDescription());
+            return [$errorMessage];
+        }
```
Comment (TomasVotruba) [inline utils/phpstan-extensions/src/Rule/CheckNotTestsNamespaceOutsideTestsDirectoryRule.php:43]: Could you refactor this to early return, so it's better readable?

File: utils/phpstan-extensions/config/phpstan-extensions.neon [hunk 2]
```diff
@@ -11,6 +11,10 @@ services:
          class: Rector\PHPStanExtensions\Rule\KeepRectorNamespaceForRectorRule
          tags: [phpstan.rules.rule]
 
+    -
+         class: Rector\PHPStanExtensions\Rule\CheckNotTestsNamespaceOutsideTestsDirectoryRule
+         tags: [phpstan.rules.rule]
```
Comment (TomasVotruba) [inline utils/phpstan-extensions/config/phpstan-extensions.neon:16]: :+1:

**Comment by dobryy** [comment]
At the moment, I have no idea how to make it smarter :)
If it fine to leave it as it is I'd do some fine-tuning for if condition and remove the `TestNameSpace` and it is ready for merging.

**Comment by TomasVotruba** [comment]
> Also the test is failing because the test fixture is in the tests directory of the phpstan-extensions util.

Make it pass without no error, as the location correct (there is namespace Tests and location tests)

**Comment by TomasVotruba** [comment]
Looks good, thanks :+1:

File: utils/phpstan-extensions/tests/Rule/CheckNotTestsNamespaceOutsideTestsDirectoryRule/Fixture/tests/TestsNamespaceOutsideTestsDirectoryClass.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+
+namespace Rector\PHPStanExtensions\Tests\Rule\CheckNotTestsNamespaceOutsideTestsDirectoryRule\Fixture\tests;
```
Comment (TomasVotruba) [inline utils/phpstan-extensions/tests/Rule/CheckNotTestsNamespaceOutsideTestsDirectoryRule/Fixture/tests/TestsNamespaceOutsideTestsDirectoryClass.php:4]: I just see it, this should be 

```diff
-namespace Rector\PHPStanExtensions\Tests\Rule\CheckNotTestsNamespaceOutsideTestsDirectoryRule\Fixture\tests;
+namespace Rector\PHPStanExtensions\Tests\Rule\CheckNotTestsNamespaceOutsideTestsDirectoryRule\Fixture\Tests;
```

Comment (TomasVotruba) [reply utils/phpstan-extensions/tests/Rule/CheckNotTestsNamespaceOutsideTestsDirectoryRule/Fixture/tests/TestsNamespaceOutsideTestsDirectoryClass.php:4]: Without that, the class is not autoloaded.

**Pull request merged**